src/distances.jl

```diff
@@ -24,64 +84,45 @@ julia> Ellipsoidal([1.0,0.5], [π/2])
 julia> Ellipsoidal([1.0,0.5,0.5], [π/2,0.0,0.0])
 ```
 """
-struct Ellipsoidal{N,T} <: Metric
-  dist::Mahalanobis{T}
 
-  function Ellipsoidal{N,T}(semiaxes, angles) where {N,T}
+struct Ellipsoidal{N,M<:Mahalanobis} <: Metric
+  dist::M
+
+  function Ellipsoidal{N,T}(semiaxes, angles; convention=:TaitBryanExtr) where {N,T}
     @assert length(semiaxes) == N "number of semiaxes must match spatial dimension"
     @assert all(semiaxes .> zero(T)) "semiaxes must be positive"
     @assert N ∈ [2,3] "dimension must be either 2 or 3"
 
-    # scaling matrix
-    Λ = Diagonal(one(T)./semiaxes.^2)
-
-    # rotation matrix
-    if N == 2
-      θ = angles[1]
-
-      cosθ = cos(θ)
-      sinθ = sin(θ)
+    rule = rules[convention]
 
-      P = [cosθ -sinθ
-           sinθ  cosθ]
+    # invert x and y if necessary
+    if rule.main == :y
+       semiaxes[1], semiaxes[2] = semiaxes[2], semiaxes[1]
     end
-    if N == 3
-      θxy, θyz, θzx = angles
-
-      cosxy = cos(θxy)
-      sinxy = sin(θxy)
-      cosyz = cos(θyz)
-      sinyz = sin(θyz)
-      coszx = cos(θzx)
-      sinzx = sin(θzx)
-
-      _1 = one(T)
-      _0 = zero(T)
-
-      Rxy = [cosxy -sinxy _0
-             sinxy  cosxy _0
-             _0     _0 _1]
 
-      Ryz = [_1    _0     _0
-             _0 cosyz -sinyz
-             _0 sinyz  cosyz]
+    # scaling matrix
+    Λ = Diagonal(SVector{N}(one(T)./semiaxes.^2))
 
-      Rzx = [ coszx _0 sinzx
-             _0 _1    _0
-             -sinzx _0 coszx]
+    # convert to radian and invert sign if necessary
+    !rule.radian && (angles = deg2rad.(angles))
+    _0 = zero(eltype(angles))
+    N == 2 && (angles = [angles[1], _0, _0])
+    intr = @. (rule.motion == :CW) & !rule.extrinsic
+    extr = @. (rule.motion == :CCW) & rule.extrinsic
+    angles[intr .| extr] *= -1
 
-      P = Rzx*Ryz*Rxy
-    end
+    # rotation matrix
+    P = angle_to_dcm(angles..., rule.order)[SOneTo(N),SOneTo(N)]
 
     # ellipsoid matrix
-    Q = P*Λ*P'
+    Q = rule.extrinsic ? P*Λ*P' : P'*Λ*P
 
     new(Mahalanobis(Q))
   end
 end
 
-Ellipsoidal(semiaxes::AbstractVector{T}, angles::AbstractVector{T}) where {T} =
-  Ellipsoidal{length(semiaxes),T}(semiaxes, angles)
+Ellipsoidal(semiaxes::AbstractVector{T}, angles::AbstractVector{T}; kwargs...) where {T} =
+  Ellipsoidal{length(semiaxes),T}(semiaxes, angles; kwargs...)
```

Comment:
Also this constructor needs an update after the new type parameter M

Reply:
Could make it work with some extra changes too. Is that ok?

```julia
Ellipsoidal(semiaxes::AbstractVector, angles::AbstractVector; kwargs...) =
  Ellipsoidal{length(semiaxes),Mahalanobis}(semiaxes, angles; kwargs...)

evaluate(dist::Ellipsoidal{N,M}, a::AbstractVector, b::AbstractVector) where {N,M} =
  evaluate(dist.dist, a, b)
```